cmd/sourced/cmd/web.go

```diff
@@ -32,54 +33,177 @@ func init() {
 	rootCmd.AddCommand(&webCmd{})
 }
 
-func openUI() error {
-	var stdout bytes.Buffer
-	// wait for the container to start, it can take a while in some cases
-	for {
-		err := compose.RunWithIO(context.Background(),
-			os.Stdin, &stdout, nil, "port", containerName, "8088")
+func openUI(address string) error {
+	// docker-compose returns 0.0.0.0 which is correct for the bind address
+	// but incorrect as connect address
+	url := fmt.Sprintf("http://%s", strings.Replace(address, "0.0.0.0", "127.0.0.1", 1))
 
-		if err == nil {
+	for {
+		client := http.Client{Timeout: time.Second}
+		if _, err := client.Get(url); err == nil {
 			break
 		}
 
-		if workdir.ErrMalformed.Is(err) || dir.ErrNotExist.Is(err) || dir.ErrNotValid.Is(err) {
-			return err
-		}
-
 		time.Sleep(1 * time.Second)
 	}
 
-	address := strings.TrimSpace(stdout.String())
-	if address == "" {
-		return fmt.Errorf("could not find the public port of %s", containerName)
+	if err := browser.OpenURL(url); err != nil {
+		return errors.Wrap(err, "could not open the browser")
 	}
 
-	// docker-compose returns 0.0.0.0 which is correct for the bind address
-	// but incorrect as connect address
-	url := fmt.Sprintf("http://%s", strings.Replace(address, "0.0.0.0", "127.0.0.1", 1))
+	return nil
+}
+
+func checkFailFast(stdout *bytes.Buffer) (bool, error) {
+	err := compose.RunWithIO(context.Background(),
+		os.Stdin, stdout, nil, "port", containerName, "8088")
+	if workdir.ErrMalformed.Is(err) || dir.ErrNotExist.Is(err) || dir.ErrNotValid.Is(err) {
+		return true, err
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	return false, nil
+}
 
+func waitForContainer(stdout *bytes.Buffer) {
 	for {
-		client := http.Client{Timeout: time.Second}
-		if _, err := client.Get(url); err == nil {
+		if err := compose.RunWithIO(context.Background(),
+			os.Stdin, stdout, nil, "port", containerName, "8088"); err == nil {
 			break
 		}
 
 		time.Sleep(1 * time.Second)
 	}
+}
 
-	if err := browser.OpenURL(url); err != nil {
-		return errors.Wrap(err, "could not open the browser")
+var newLineFormatter = regexp.MustCompile(`(\r\n|\r|\n)`)
+var stateExtractor = regexp.MustCompile(`(?m)^srcd-[\w\d]+.*(Up|Exit (\w+))`)
```

Comment:
should it be `(Exit (\d+)`?

Reply:
Ops, nice catch! And also the `\d` is redundant in the first part, in the end it will be as follows:
```
(?m)^srcd-\w+.*(Up|Exit (\d+))
```

Reply:
Done [here](https://github.com/src-d/sourced-ce/pull/186/commits/2a9ce5425fe437c68c458a0beec0818d61c4d387).